File: 230.二叉搜索树中第k小的元素.go
/*
 * @lc app=leetcode.cn id=230 lang=golang
 *
 * [230] 二叉搜索树中第K小的元素
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
package main
func kthSmallest(root *TreeNode, k int) int {
	stack := []*TreeNode{}
	var ans int
	for root != nil || len(stack) != 0 {//借用了树的遍历，这里是前序
		for root != nil {
			stack = append(stack,root)
			root = root.Left
		}
		node := stack[len(stack)-1]
		k--
		if k == 0 {
			ans = node.Val
		}
		stack = stack[:len(stack)-1]
		root = node.Right
	}
	return ans
}
// @lc code=end

File: 230.二叉搜索树中第-k-小的元素.ts
/*
 * @lc app=leetcode.cn id=230 lang=typescript
 *
 * [230] 二叉搜索树中第 K 小的元素
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
//  中序遍历
function kthSmallest(root: TreeNode | null, k: number): number {
    let q: TreeNode[] = new Array<TreeNode>();
    let res: number = 0;
    for (;root !== null || q.length !==0;) {
        for (;root != null;) {
            q.push(root);
            root = root.left;
        }
        let node = q.pop();
        k--;
        if (k === 0) {
            return node.val;
        }
        root = node.right;
    }
};
// @lc code=end

